clippy_lints/src/methods/swap_with_temporary.rs

```diff
@@ -0,0 +1,142 @@
+        |diag| {
+            const DROP_MSG: &str = "drop them if creating these temporary expressions is necessary";
+
+            diag.span_note(left_temp.span, MSG_TEMPORARY);
+            diag.span_note(right_temp.span, MSG_TEMPORARY);
+
+            // If the `swap()` is a statement by itself, just propose to replace `swap(&mut a, &mut b)` by `a;
+            // b;` in order to drop `a` and `b` while acknowledging their side effects. If the
+            // `swap()` call is part of a larger expression, replace it by `{core,
+            // std}::mem::drop((a, b))`.
+            if matches!(cx.tcx.parent_hir_node(expr.hir_id), Node::Stmt(..)) {
+                diag.multipart_suggestion(
+                    DROP_MSG,
+                    vec![
+                        (func.span.with_hi(left_temp.span.lo()), String::new()),
+                        (left_temp.span.between(right_temp.span), String::from("; ")),
+                        (expr.span.with_lo(right_temp.span.hi()), String::new()),
+                    ],
+                    Applicability::MachineApplicable,
+                );
+            } else {
+                diag.multipart_suggestion(
+                    DROP_MSG,
+                    vec![
+                        (
+                            func.span,
+                            format!("{}::mem::drop(", if is_no_std_crate(cx) { "core" } else { "std" }),
+                        ),
+                        (left.span.with_hi(left_temp.span.lo()), String::new()),
+                        (right.span.with_hi(right_temp.span.lo()), String::new()),
+                        (expr.span.shrink_to_hi(), String::from(")")),
+                    ],
+                    Applicability::MachineApplicable,
+                );
+            }
+        },
+    );
+}
+
+fn emit_lint_assign(cx: &LateContext<'_>, expr: &Expr<'_>, target: &ArgKind<'_>, temp: &Expr<'_>) {
+    span_lint_and_then(
+        cx,
+        SWAP_WITH_TEMPORARY,
+        expr.span,
+        "swapping with a temporary value is inefficient",
+        |diag| {
+            diag.span_note(temp.span, MSG_TEMPORARY);
+            let mut applicability = Applicability::MachineApplicable;
+            let ctxt = expr.span.ctxt();
+            let assign_target = match target {
+                ArgKind::Expr(target) | ArgKind::RefMutToPlaceAsMacro(target) => {
+                    Sugg::hir_with_context(cx, target, ctxt, "_", &mut applicability).deref()
+                },
+                ArgKind::RefMutToPlace(target) => Sugg::hir_with_context(cx, target, ctxt, "_", &mut applicability),
+                ArgKind::RefMutToTemp(_) => unreachable!(),
+            };
+            let assign_source = Sugg::hir_with_context(cx, temp, ctxt, "_", &mut applicability);
+
+            // If the `swap()` is a statement by itself, propose to replace it by `a = b`. Otherwise, when part
+            // of a larger expression, surround the assignment with a block to make it `()`.
```

Comment:
Assignment expressions also always evaluate to `()` so I don't *think* wrapping it in a block is needed?

Reply:
Removed, as we only consider the statement and latest expression in a block when autofixing. Other cases require manual examination, as they have probably been written for a reason.